File: print-service/src-tauri/src/pos_quotation_ticket_escpos.rs
//! Ticket de cotización 80 mm en ESC/POS (alineado a `quotation-receipt-print.ts`).

use crate::pos_quotation_ticket::{parse_pos_quotation_ticket_from_value, PosQuotationTicket, QuotationLine};
use crate::pos_sale_ticket_escpos::{
    append_barcode_centered, append_divider, append_label_value_wrapped, append_line,
    append_product_line_block, append_section_gap, append_ticket_logo, escpos_align,
    escpos_apply_ticket_typography,
    escpos_bold, escpos_dense_body, escpos_double_height_off, escpos_double_height_on, escpos_init,
    format_clp, format_datetime, money, pad_label_value, pad_left, wrap_lines, WIDTH,
};
use anyhow::Result;
use std::path::PathBuf;

fn line_unit_price_with_tax(line: &QuotationLine) -> f64 {
    if line.quantity.abs() > 0.001 {
        line.total / line.quantity
    } else {
        line.unit_price
    }
}

fn line_display_name(line: &QuotationLine) -> String {
    let base = line.product_name.trim();
    let variant = line.variant_name.as_deref().unwrap_or("").trim();
    let mut name = if variant.is_empty() {
        base.to_string()
    } else if base.is_empty() {
        variant.to_string()
    } else {
        format!("{base} · {variant}")
    };
    if let Some(sku) = line.product_sku.as_deref().map(str::trim).filter(|s| !s.is_empty()) {
        if !name.is_empty() {
            name.push_str(&format!(" ({sku})"));
        } else {
            name = sku.to_string();
        }
    }
    name
}

fn append_quotation_totals(buf: &mut Vec<u8>, q: &PosQuotationTicket) {
    append_line(buf, &pad_left("Subtotal:", &money(q.subtotal)));
    append_line(buf, &pad_left("Impuestos:", &money(q.tax_amount)));
    if q.discount_amount > 0.01 {
        append_line(
            buf,
            &pad_left("Descuentos:", &format!("-${}", format_clp(q.discount_amount))),
        );
    }
    escpos_bold(buf, true);
    append_line(buf, &pad_left("TOTAL:", &money(q.total)));
    escpos_bold(buf, false);
    append_divider(buf);
}

fn append_wrapped_section(buf: &mut Vec<u8>, title: &str, body: &str) {
    let body = body.trim();
    if body.is_empty() {
        return;
    }
    append_divider(buf);
    escpos_bold(buf, true);
    append_line(buf, title);
    escpos_bold(buf, false);
    for line in wrap_lines(body, WIDTH) {
        append_line(buf, &line);
    }
}

pub fn build_pos_quotation_ticket_escpos(q: &PosQuotationTicket) -> Result<Vec<u8>> {
    let mut buf = escpos_init();
    escpos_apply_ticket_typography(&mut buf);

    append_ticket_logo(&mut buf, q.company.logo_base64.as_deref());

    let store = q
        .company
        .nombre_fantasia
        .as_deref()
        .filter(|s| !s.trim().is_empty())
        .unwrap_or(q.company.razon_social.as_str());

    escpos_align(&mut buf, 1);
    escpos_bold(&mut buf, true);
    escpos_double_height_on(&mut buf);
    for line in wrap_lines(store, WIDTH / 2) {
        append_line(&mut buf, &line);
    }
    escpos_double_height_off(&mut buf);
    escpos_bold(&mut buf, false);

    if let Some(fantasy) = q.company.nombre_fantasia.as_deref() {
        let rs = q.company.razon_social.trim();
        if !rs.is_empty() && fantasy.trim() != rs {
            for line in wrap_lines(rs, WIDTH) {
                append_line(&mut buf, &line);
            }
        }
    }
    if let Some(rut) = q.company.rut.as_deref().filter(|s| !s.trim().is_empty()) {
        append_line(&mut buf, &format!("RUT: {}", rut.trim()));
    }
    if let Some(act) = q
        .company
        .business_activity
        .as_deref()
        .filter(|s| !s.trim().is_empty())
    {
        for line in wrap_lines(act.trim(), WIDTH) {
            append_line(&mut buf, &line);
        }
    }

    append_divider(&mut buf);
    escpos_bold(&mut buf, true);
    append_line(&mut buf, "COTIZACION");
    escpos_bold(&mut buf, false);

    let folio = q.document_number.trim();
    append_line(&mut buf, &format!("Emitida: {}", format_datetime(&q.issued_at)));
    append_line(
        &mut buf,
        &format!("Valida hasta: {}", format_datetime(&q.valid_until)),
    );

    if let Some(b) = q.branch_name.as_deref().filter(|s| !s.trim().is_empty()) {
        append_line(&mut buf, &pad_left("Sucursal:", b.trim()));
    }
    if let Some(p) = q.point_of_sale_name.as_deref().filter(|s| !s.trim().is_empty()) {
        append_line(&mut buf, &pad_left("Punto venta:", p.trim()));
    }

    let has_customer = q.customer_name.as_deref().map(|s| !s.trim().is_empty()).unwrap_or(false)
        || q
            .customer_document
            .as_deref()
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false);
    if has_customer {
        append_divider(&mut buf);
        append_line(&mut buf, "Cliente");
        if let Some(name) = q.customer_name.as_deref().filter(|s| !s.trim().is_empty()) {
            append_label_value_wrapped(&mut buf, "Nombre:", name.trim());
        }
        if let Some(doc) = q.customer_document.as_deref().filter(|s| !s.trim().is_empty()) {
            append_line(&mut buf, &pad_label_value("Documento:", doc.trim()));
        }
    }

    append_divider(&mut buf);
    escpos_dense_body(&mut buf);
    escpos_align(&mut buf, 1);
    escpos_bold(&mut buf, true);
    append_line(&mut buf, "DETALLE");
    escpos_bold(&mut buf, false);
    escpos_align(&mut buf, 0);
    append_section_gap(&mut buf);

    for (idx, line) in q.lines.iter().enumerate() {
        let name = line_display_name(line);
        let qty = if (line.quantity.fract()).abs() < 0.001 {
            format!("{}", line.quantity.round() as i64)
        } else {
            format!("{:.2}", line.quantity)
        };
        let qty_unit = format!("{qty}x {}", money(line_unit_price_with_tax(line)));
        append_product_line_block(&mut buf, &name, &qty_unit, &money(line.total));
        if idx + 1 < q.lines.len() {
            append_section_gap(&mut buf);
        }
    }

    append_divider(&mut buf);
    append_quotation_totals(&mut buf, q);

    if let Some(notes) = q.notes.as_deref() {
        append_wrapped_section(&mut buf, "Notas", notes);
    }
    if let Some(terms) = q.terms.as_deref() {
        append_wrapped_section(&mut buf, "Condiciones", terms);
    }

    if !folio.is_empty() {
        append_barcode_centered(&mut buf, folio);
    }

    let footer = format!("{folio} - {}", format_datetime(&q.issued_at));
    escpos_align(&mut buf, 1);
    append_line(&mut buf, &footer);
    escpos_align(&mut buf, 0);
    append_line(&mut buf, "");

    Ok(buf)
}

pub fn write_pos_quotation_ticket_escpos_from_value(
    dir: &PathBuf,
    value: &serde_json::Value,
) -> Result<PathBuf> {
    std::fs::create_dir_all(dir)?;
    let q = parse_pos_quotation_ticket_from_value(value)?;
    let bytes = build_pos_quotation_ticket_escpos(&q)?;
    let id = uuid::Uuid::new_v4().to_string();
    let p = dir.join(format!("quotation_ticket_{id}.escpos"));
    std::fs::write(&p, &bytes)?;
    Ok(p)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn quotation_escpos_has_cotizacion_heading() {
        let v = serde_json::json!({
            "documentNumber": "COT-001",
            "issuedAt": "2026-01-01T12:00:00Z",
            "validUntil": "2026-02-01T12:00:00Z",
            "company": { "razonSocial": "Tienda" },
            "lines": [{
                "productName": "Item",
                "quantity": 1,
                "unitPrice": 1000,
                "total": 1190
            }],
            "subtotal": 1000,
            "taxAmount": 190,
            "total": 1190
        });
        let q = parse_pos_quotation_ticket_from_value(&v).unwrap();
        let bytes = build_pos_quotation_ticket_escpos(&q).unwrap();
        assert!(bytes.windows(10).any(|w| w == b"COTIZACION"));
    }
}

File: kai-printers-android/app/src/main/java/com/kaistore/printers/print/PosQuotationTicketEscPos.kt
package com.kaistore.printers.print

import kotlinx.serialization.json.Json

object PosQuotationTicketEscPos {
    fun fromTicketJson(ticketJson: String, widthChars: Int = 48): ByteArray {
        val q = Json.parseToJsonElement(ticketJson).jsonObj()
            ?: throw IllegalStateException("invalid_ticket_json")
        val w = EscPosWriter(widthChars)
        w.beginTicket()

        val company = q.jsonObj("company")
        EscPosLogo.appendIfPresent(w, company?.jsonStr("logoBase64"))
        w.appendStoreHeader(company)

        w.divider()
        w.bold(true)
        w.line("COTIZACION")
        w.bold(false)

        val folio = q.jsonStr("documentNumber").present().orEmpty()
        w.line("Emitida: ${w.formatDateTime(q.jsonStr("issuedAt").orEmpty())}")
        w.line("Valida hasta: ${w.formatDateTime(q.jsonStr("validUntil").orEmpty())}")
        q.jsonStr("branchName").present()?.let { w.line(w.padLeft("Sucursal:", it)) }
        q.jsonStr("pointOfSaleName").present()?.let { w.line(w.padLeft("Punto venta:", it)) }

        val customerName = q.jsonStr("customerName").present()
        val customerDoc = q.jsonStr("customerDocument").present()
        if (customerName != null || customerDoc != null) {
            w.divider()
            w.line("Cliente")
            customerName?.let { w.labelValueWrapped("Nombre:", it) }
            customerDoc?.let { w.line(w.padLabelValue("Documento:", it)) }
        }

        w.divider()
        w.alignCenter(true)
        w.bold(true)
        w.line("DETALLE")
        w.bold(false)
        w.alignCenter(false)
        w.sectionGap()

        val lines = q.jsonArr("lines").orEmpty()
        lines.forEachIndexed { idx, row ->
            val obj = row.jsonObj() ?: return@forEachIndexed
            val name = quotationLineName(obj)
            val qty = obj.jsonNum("quantity") ?: 0.0
            val total = obj.jsonNum("total") ?: 0.0
            val unitPrice = obj.jsonNum("unitPrice") ?: if (qty > 0) total / qty else 0.0
            val qtyStr = formatQty(qty)
            val qtyUnit = "${qtyStr}x ${w.money(unitPrice)}"
            w.appendProductLineBlock(name, qtyUnit, w.money(total))
            if (idx + 1 < lines.size) w.sectionGap()
        }

        w.divider()
        w.line(w.padLeft("Subtotal:", w.money(q.jsonNum("subtotal") ?: 0.0)))
        w.line(w.padLeft("Impuestos:", w.money(q.jsonNum("taxAmount") ?: 0.0)))
        val discount = q.jsonNum("discountAmount") ?: 0.0
        if (discount > 0.01) {
            w.line(w.padLeft("Descuentos:", "-${w.money(discount)}"))
        }
        w.bold(true)
        w.line(w.padLeft("TOTAL:", w.money(q.jsonNum("total") ?: 0.0)))
        w.bold(false)
        w.divider()

        w.appendWrappedSection("Notas", q.jsonStr("notes"))
        w.appendWrappedSection("Condiciones", q.jsonStr("terms"))

        if (folio.isNotEmpty()) w.appendBarcodeCentered(folio)
        w.alignCenter(true)
        w.line("$folio - ${w.formatDateTime(q.jsonStr("issuedAt").orEmpty())}")
        w.alignCenter(false)
        w.line()

        return w.toByteArray(openDrawer = false)
    }

    private fun quotationLineName(obj: kotlinx.serialization.json.JsonObject): String {
        val base = obj.jsonStr("productName").present().orEmpty()
        val variant = obj.jsonStr("variantName").present().orEmpty()
        var name = when {
            variant.isEmpty() -> base
            base.isEmpty() -> variant
            else -> "$base · $variant"
        }
        val sku = obj.jsonStr("productSku").present()
        if (sku != null) {
            name = if (name.isNotEmpty()) "$name ($sku)" else sku
        }
        return name
    }

    private fun formatQty(qty: Double): String =
        if (kotlin.math.abs(qty % 1) < 0.001) qty.toLong().toString()
        else String.format(java.util.Locale.US, "%.2f", qty)
}
